web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts

```diff
@@ -3,52 +3,278 @@ import { createSlice, PayloadAction } from '@reduxjs/toolkit'
+import { NUM_WEEK_DAYS } from 'features/hfiCalculator/constants'
+import { FireCentre } from 'api/hfiCalcAPI'
+import { DateTime } from 'luxon'
 
-interface State {
+export interface HFIResult {
+  dailies: StationDaily[]
+  dailyMeanIntensity: number | undefined
+  dailyMeanIntensityGroups: (number | undefined)[]
+  maxMeanIntensityGroup: number | undefined
+  dailyPrepLevel: number | undefined
+  dailyPrepLevels: (number | undefined)[]
+  meanPrepLevel: number | undefined
+}
+
+export interface HFICalculatorState {
   loading: boolean
   error: string | null
   dailies: StationDaily[]
+  fireCentres: { [key: string]: FireCentre }
+  numPrepDays: number
+  selected: number[]
+  selectedPrepDate: string
+  formattedDateStringHeaders: string[]
+  planningAreaHFIResults: { [key: string]: HFIResult }
 }
 
-const initialState: State = {
+const initialState: HFICalculatorState = {
   loading: false,
   error: null,
-  dailies: []
+  dailies: [],
+  fireCentres: {},
+  numPrepDays: NUM_WEEK_DAYS,
+  selected: [],
+  selectedPrepDate: '',
+  formattedDateStringHeaders: [],
+  planningAreaHFIResults: {}
+}
+
+export const calculateMeanIntensity = (dailies: StationDaily[]): number | undefined =>
+  dailies.length === 0
+    ? undefined
+    : Math.round(
+        (10 * dailies.map(daily => daily.intensity_group).reduce((a, b) => a + b, 0)) /
+          dailies.length
+      ) / 10
+
+export const calculateDailyMeanIntensities = (
+  dailies: StationDaily[],
+  numPrepDays: number
+): (number | undefined)[] => {
+  const utcDict = groupBy(dailies, (daily: StationDaily) => daily.date.toUTC().toMillis())
+
+  const dailiesByDayUTC = new Map(
+    Object.entries(utcDict).map(entry => [Number(entry[0]), entry[1]])
+  )
+
+  return range(numPrepDays).map(i => {
+    const orderedDayTimestamps = Array.from(dailiesByDayUTC.keys()).sort((a, b) => a - b)
+
+    const dailiesForDay: StationDaily[] | undefined = dailiesByDayUTC.get(
+      orderedDayTimestamps[i]
+    )
+    return dailiesForDay ? calculateMeanIntensity(dailiesForDay) : undefined
+  })
+}
+
+export const calculateMaxMeanIntensityGroup = (
+  dailyMeanIntensityGroups: (number | undefined)[]
+): number => Math.max(...dailyMeanIntensityGroups.filter(isDefined))
+
+const isDefined = (item: number | undefined): item is number => {
+  return !!item
+}
+
+export const calculateDailyPrepLevels = (
+  dailyMeanIntensityGroups: (number | undefined)[]
+): (number | undefined)[] => {
+  const prepLevels: (number | undefined)[] = []
+  range(NUM_WEEK_DAYS).map(day => {
+    const meanIntensityGroup = dailyMeanIntensityGroups[day]
+    prepLevels.push(calculatePrepLevel(meanIntensityGroup))
+  })
+  return prepLevels
+}
+
+export const calculatePrepLevel = (
+  meanIntensityGroup: number | undefined
+): number | undefined => {
+  // for now, prep level calculation assumed a fixed Fire Starts value of 0-1
+
+  if (isUndefined(meanIntensityGroup)) {
+    return undefined
+  } else {
```

Comment:
pedantry - the `else` statement isn't required because the `if` has a return statement